add support for pcap-ng

mccanne opened the pull request:
This commit adds support for pcap-ng.  The pcapio.Reader interface
was tweaked a bit to accommodate the changes so that a Reader can
be either a pcap-ng reader or a legacy pcap reader.  pcapio.Reader
uses the same trick as zio to autodetect whether input is pcap-ng
or pcap legacy.

The pcap-ng parsing code is derived from
https://github.com/google/gopacket/.

We added a simple test for pcap-ng along the lines of the pcap test
that was already there.  Testing is still limited for pcap and
needs to be addressed in a future PR.

henridf commented:
I noticed this earlier and wanted to dig in to the PR and `ranger` to understand things but didn't get around to it, so... some very black-boxy feedback:

It looks like result of `pcap index` on a (legacy) pcap file have changed with this PR.  Is this expected?

Here's master:
```json
    "Index": [
      {
        "X": 24,
        "Y0": 0,
        "Y1": 1521832548719396000
      },
      {
        "X": 390608,
        "Y0": 1521832548719400000,
        "Y1": 1521832548739203000
      },
      {
        "X": 799324,
        "Y0": 1521832548739205000,
        "Y1": 1521832548755799000
      },
      {
        "X": 1114496,
        "Y0": 1521832548755801000,
        "Y1": 1521832548765437000
      },
      {
```

and on this branch:
```json
    "Index": [
      {
        "X": 24,
        "Y0": 1521832548699327000,
        "Y1": 1521832548719400000
      },
      {
        "X": 390684,
        "Y0": 1521832548719488000,
        "Y1": 1521832548739205000
      },
      {
        "X": 799400,
        "Y0": 1521832548739251000,
        "Y1": 1521832548755801000
      },
   ```

mccanne commented:
> It looks like result of `pcap index` on a (legacy) pcap file have changed with this PR. Is this expected?

I didn't expect that but I'm not surprised, so I'll have a look.  It looks like it still is correct, but I will double check.  Probably won't get to this for a few hours.

I changed a lot and found and fixed several bugs so I'll try to do some more comprehensive QAing...

alfred-landrum commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,317 @@
+			packet := block[28:]
+			if len(packet) < caplen {
+				return nil, 0, ErrCorruptPcap
+			}
```
There's a lot of commonality between this and `Packet()`; would it be possible to share the logic?

mccanne replied:
good idea

alfred-landrum commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,317 @@
+
+// Resolution returns the timestamp resolution of acquired timestamps before scaling to NanosecondTimestampResolution.
+func (r *NgReader) Resolution() gopacket.TimestampResolution {
+	//XXX
```
If this isn't used, would you remove the method?

mccanne replied:
ok

mccanne commented:
> > It looks like result of `pcap index` on a (legacy) pcap file have changed with this PR. Is this expected?
> 
> I didn't expect that but I'm not surprised, so I'll have a look. It looks like it still is correct, but I will double check. Probably won't get to this for a few hours.
> 
> I changed a lot and found and fixed several bugs so I'll try to do some more comprehensive QAing...

Ok @henridf, I looked at this a bit closer and I think the newer code handles the first point correctly with the re-write to handle pcap-ng.  The first index entry should be X=24 (the size of the pcap legacy header and the Y0/Y1 range should be the timestamp range, which should *no*t start from 0 as with the old code in master.

mccanne commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,317 @@
+}
+
+func (r *NgReader) Packet(block []byte) ([]byte, nano.Ts, layers.LinkType) {
+	//XXX support only enhance packet blocks right now
```
will add clarifying comment for the Packet() method

mccanne commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,317 @@
+var ErrCorruptPcap = errors.New("bad pcap file")
+
+// NgReaderOptions holds options for reading a pcapng file
+type NgReaderOptions struct {
```
ok, that was the gopacket code.

mccanne commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,317 @@
+
+var ErrCorruptPcap = errors.New("bad pcap file")
+
+// NgReaderOptions holds options for reading a pcapng file
```
ok, gopacket code

mccanne commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,317 @@
+
+// NgReaderOptions holds options for reading a pcapng file
+type NgReaderOptions struct {
+	// ErrorOnMismatchingLinkType enables returning an error if a packet with a link type not matching the first interface is encountered and WantMixedLinkType == false.
```
now that I'm making these edits, we doon't have a use for these options so I will just delete all thsi stuff.

mccanne commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,317 @@
+	if ifno >= len(r.ifaces) {
+		return nil, 0, 0
+	}
+	t := time.Unix(r.convertTime(ifno, uint64(r.getUint32(block[12:16]))<<32|uint64(r.getUint32(block[16:20])))).UTC()
```
Ok, will do.  I think this can be simplified but didn't want to break the existing pcapgo code.

mccanne commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,317 @@
+	return nil
+}
+
+// readOption reads a single arbitrary option (type and value). If there is no space left for options and end of options is missing, it is faked.
```
will fix

mccanne commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,317 @@
+	}
+
+	//parse options
+	if intf.TimestampResolution.Binary() {
```
I'd rather just leave it for now since it's the pcapgo code.

mattnibs commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,310 @@
+}
+
+// Packet returns the captured portion of a packet from an enhanced packet
+// block return by Read() (i.e., with BlockTYpe equal to TypePacket) beginning
```
```suggestion
// block return by Read() (i.e., with BlockType equal to TypePacket) beginning
```

mattnibs commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,310 @@
+// with the link-layer header.  It also extracts the capture timestamp and link
+// layer type and returns those values along with the packet.  If an error is
+// encountered, zero values are returned for the three values.  PCAP-NG simple
+// packet types aren't supported yet (we presume thiis type of trace is rare and
```
```suggestion
// packet types aren't supported yet (we presume this type of trace is rare and
```

mattnibs commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,310 @@
+// encountered, zero values are returned for the three values.  PCAP-NG simple
+// packet types aren't supported yet (we presume thiis type of trace is rare and
+// does not fit our use case here as these traces do not include capture timestamps
+// annd the point here is to pull our ranges of packets from a large pcap based
```
```suggestion
// and the point here is to pull our ranges of packets from a large pcap based
```

mattnibs commented on pcap/index.go:
```diff
@@ -21,40 +21,72 @@ type Section struct {
-		return nil, err
-	}
+// CreateIndex creates an index for a pcap file.
+func CreateIndex(reader pcapio.Reader, limit int) (Index, error) {
```
Unless necessary I would be obliged to keep this interface the same- more boiler plate for clients.

mccanne replied:
excellent point.  will fix.

mattnibs commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,310 @@
+	if binary.BigEndian.Uint32(b[8:12]) == ngByteOrderMagic {
+		r.bigEndian = true
+	} else if binary.LittleEndian.Uint32(b[8:12]) == ngByteOrderMagic {
+		r.bigEndian = false
```
nit: Instead of this why not replace `bigEndian bool` with `byteOrder binary.ByteOrder` and do:
```golang
if binary.BigEndian.Uint32(b[8:12]) == ngByteOrderMagic {
	r.byteOrder = binary.BigEndian
} else if binary.LittleEndian.Uint32(b[8:12]) == ngByteOrderMagic {
	r.byteOrder = binary.LittleEndian
```
This way you can get rid of your `getUint64/32/16` helpers.

mccanne replied:
How about later?  This is all gopacket logic that I was just hoping would work for us.

mattnibs replied:
ok

mattnibs commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,310 @@
+		r.bigEndian = false
+	} else {
+		return errors.New("Wrong byte order value in Section Header")
+	}
```
nit: It would be nice for this to check Major Version and throw an error if != 1 (or at least a comment about that).

mccanne replied:
How about later? This is all gopacket logic that I was just hoping would work for us.

mattnibs replied:
ok

mattnibs commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,310 @@
+	}
+
+	//pcapng _must_ start with a section header
+	typ, block, err := ret.readBlock()
```
If this pcap was recorded with written with bigEndian byte order (the default is littleEndian) I believe this going to get return an error: "pcap-ng trailer length mismatch" (Same goes for `Read()`. I believe you're going to need to read the block type bits, and if its a section header run this through `parseSectionHeader`, before  calling `readBlock`.

mccanne replied:
Will fix it!

mccanne replied:
good catch

mattnibs commented on zqd/handlers.go:
```diff
@@ -85,6 +86,12 @@ func handlePacketSearch(c *Core, w http.ResponseWriter, r *http.Request) {
 	}
+	pcapReader, err := pcapio.NewReader(slicer)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
```
```suggestion
		http.Error(w, err.Error(), http.StatusInternalServerError)
```

mccanne replied:
I didn't think that was right but I copied it from the one above, which should also change right?

mattnibs replied:
Yes.

mattnibs commented on pcap/pcapio/reader.go:
```diff
@@ -27,6 +24,23 @@ type Info struct {
+	_, err := NewPcapReader(track)
+	if err == nil {
+		return NewPcapReader(recorder)
+	}
```
Can fixed in a follow up pr but if this returns an error because it's an unsupported pcap version, that error should get returned.

mccanne replied:
NewPcapReader() returns the error as the second arg

mattnibs replied:
If NewPcapReader returns a "this version of legacy pcap is invalid err" from a legacy pcap, the error will be over-written by whatever error NewNgReader finds.

mccanne replied:
No, if they both fail, then it's ErrCorruptPcap

mccanne replied:
I think this is ok.
GSD

mccanne commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,317 @@
+	typ := ngBlockType(r.getUint32(hdr[:4]))
+	length := r.getUint32(hdr[4:8])
+	if length == 0 {
+		// avoid infinite loop for bad input
```
You know, I like to point this one out because its a common flaw in protocol design and protocol implementation and thus I think it's worth it to include the comment.   May I reminisce?  I put up a bug fix at Riverbed in the CIFS protocol accelerator module because a length==0 check like this was missing and Kartik closed it with no fix explaining that a sender shouldn't put 0 in the length field.  Ugh.  Van Jacobson taught me years ago that length fields should never include the length of the fixed size header contained within referenced payload so that: (1) you don't have to check for length==0 to avoid an infinite loop (i.e., you can't make this mistake even if you tried) and (2) it lets the payload length be a little longer for the bit budget you have for the length field (which in the old days of TCP/IP header design made some difference but nowadays now big deal).  More to the point, the previous code from gopacket has not quite this bug, but the corollary, where it is computing a length incorrectly and causing an unsigned integer underflow for length value < 8.  The moral of the story is that the  block length field in the pcap-ng spec should have been defined to be the actual length less 12 (to account for the fixed size of the 4-byte block type, block length, and block length trailer) rather than defined as encoding the overall length.  It would be fun to create a pcap-ng file that is fully compliant except for a block that has a length of 0 and feed it to various pcap-related programs and see which ones lock up.  That all said, this discussion made me realize there is a different bug here, which is panic if the length is 1, 2, or 3, but according to Kartik, a file should never have those values here.  Ha!  I will fix this.

mccanne commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,317 @@
+	}
+	// The next part is a bit messed up since a section header could change the endianess...
+	// So first read then length just into a buffer, check if its a section header and then do the endianess part...
+	//XXX move this above
```
This is more gopacket vestige.  I will have a look.

mccanne replied:
oops, removed.

mccanne replied:
btw, there was profanity here from gopacket that I already edited to "messed up"

mccanne commented on cmd/pcap/cut/command.go:
```diff
@@ -130,19 +130,24 @@ func (c *Command) Run(args []string) error {
 
-	// XXX assumes legacy pcap format, will fix in next PR
-	reader, err := pcapio.NewPcapReader(in)
+	reader, err := pcapio.NewReader(in)
```
I will give it a go.

mccanne commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,317 @@
+	} else {
+		intf.scaleDown = intf.secondMask / 1e9
+	}
+	//XXX this list should be per section no?
```
good catch on the note to self.  I needed to reset the interface list in parseSectionHeader()... fixed.

mccanne commented on pcap/pcapio/ngread.go:
```diff
@@ -0,0 +1,317 @@
+}
+
+// NewNgReader initializes a new writer, reads the first section header, and if necessary according to the options the first interface.
+func NewNgReader(r io.Reader) (*NgReader, error) {
```
Yes good point, NgReaderOptions is gone now
